datastore/shared/postgresql_backend/pg_connection_handler.py

```diff
@@ -128,30 +129,34 @@ def get_connection_params(self):
         }
 
     def get_connection(self):
-        if old_conn := self.get_current_connection():
-            if old_conn.closed:
-                # If an error happens while returning the connection to the pool, it
-                # might still be set as the current connection although it is already
-                # closed. In this case, we just discard it.
-                logger.debug(f"Discarding old connection (closed={old_conn.closed})")
-                logger.debug("This indicates a previous error, please check the logs")
-                self.put_connection(old_conn, True)
-            else:
-                raise BadCodingError(
-                    "You cannot start multiple transactions in one thread!"
-                )
-        self._semaphore.acquire()
-        connection = self.connection_pool.getconn()
-        connection.autocommit = False
-        self.set_current_connection(connection)
-        return connection
-
-    def put_connection(self, connection, has_error):
-        """
-        has_error indicated, whether to not reuse the connection.
-        If the connection encountered an error, set it to true, so
-        it will be discarded from the pool.
-        """
+        with self._lock:
```

Comment:
You should use 2 separate locks for the get_ and the put_connection to prevent raise conditions.Race condition occurs when the get has the lock, but can't require the semaphore and the put has a semaphore, but can't get the lock.